Bump go mod to 1.24 and toolchain to 1.24.3

juanluisvaladas opened the pull request.

twz123 commented on pkg/component/worker/config/loader_test.go:
```diff
@@ -163,7 +163,7 @@ func TestWatchProfile(t *testing.T) {
 	cacheDir := t.TempDir()
 
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
```
Same here with lines 166-173:

```go
		ctx, cancel := context.WithCancelCause(t.Context())
		defer cancel(nil)
		timeout := errors.New("timeout: Watcher.Watch()")
		timer := time.AfterFunc(10*time.Second, func() { cancel(timeout) })
		defer func() {
			timer.Stop()
			if errors.Is(context.Cause(ctx), timeout) {
				assert.Fail(t, "Call to Watcher.Watch() took too long, check the logs for details")
			}
		}()
```

juanluisvaladas replied:
I think this `defer cancel(nil)` is redundant, so I removed it.

twz123 commented on pkg/component/worker/nllb/reconciler_test.go:
```diff
@@ -488,7 +488,7 @@ func writeKubeconfig(t *testing.T) string {
 
 func testContext(t *testing.T) context.Context {
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
```
same as the other testContext function

juanluisvaladas replied:
Like in the other, I removed the `defer cancel(nil)`

twz123 commented on pkg/component/worker/static_pods_test.go:
```diff
@@ -236,7 +236,7 @@ func TestStaticPods_Lifecycle(t *testing.T) {
 	})
 
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
`t.Context()` can be inlined.

```suggestion
```

twz123 commented on pkg/k0scontext/timeout_test.go:
```diff
@@ -96,7 +96,7 @@ func TestWithInactivityTimeout_Cancel(t *testing.T) {
 
 	t.Run("Outer", func(t *testing.T) {
-		ctx, outerCancel := context.WithCancelCause(context.TODO())
+		ctx, outerCancel := context.WithCancelCause(t.Context())
 		t.Cleanup(func() { outerCancel(nil) })
```
```suggestion
		ctx, outerCancel := context.WithCancelCause(t.Context())
```

twz123 commented on pkg/k0scontext/timeout_test.go:
```diff
@@ -96,7 +96,7 @@ func TestWithInactivityTimeout_Cancel(t *testing.T) {
+		ctx, outerCancel := context.WithCancelCause(t.Context())
 		t.Cleanup(func() { outerCancel(nil) })
 
 		ctx, cancel, _ := k0scontext.WithInactivityTimeout(ctx, time.Hour)
```
the Cleanup on line 103 is superfluous.

twz123 commented on pkg/leaderelection/client_test.go:
```diff
@@ -68,7 +68,7 @@ func TestLeaseConfig_Client(t *testing.T) {
 func TestClient_Reacquisition(t *testing.T) {
 	fakeClient := fake.NewSimpleClientset()
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
```suggestion
	ctx, cancel := context.WithCancel(t.Context())
```

twz123 commented on pkg/leaderelection/client_test.go:
```diff
@@ -142,7 +142,7 @@ func TestClient_Reacquisition(t *testing.T) {
 func TestClient_LeadTakeover(t *testing.T) {
 	fakeClient := fake.NewSimpleClientset()
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
```suggestion
	ctx, cancel := context.WithCancel(t.Context())
```

twz123 commented on pkg/leaderelection/lease_pool_test.go:
```diff
@@ -47,7 +47,7 @@ func TestLeasePoolWatcherTriggersOnLeaseAcquisition(t *testing.T) {
 	}
 
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
`t.Context()` can be inlined.

```suggestion
	ctx, cancel := context.WithCancel(t.Context())
	t.Cleanup(cancel)
```

juanluisvaladas replied:
For this one I just inlined t.Context() directly. I don't think we need the context.WithCancel because the. cancel function is only in t.Cleanup

twz123 commented on pkg/leaderelection/lease_pool_test.go:
```diff
@@ -87,7 +87,7 @@ func TestLeasePoolTriggersLostLeaseWhenCancelled(t *testing.T) {
 	}
 
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
```suggestion
	ctx, cancel := context.WithCancel(t.Context())
```

twz123 commented on pkg/leaderelection/lease_pool_test.go:
```diff
@@ -130,7 +130,7 @@ func TestLeasePoolWatcherReacquiresLostLease(t *testing.T) {
 
 	givenLeaderElectorError(nil)
-	ctx, cancel := context.WithCancel(context.TODO())
+	ctx, cancel := context.WithCancel(t.Context())
 	t.Cleanup(cancel)
```
`t.Context()` can be inlined.

```suggestion
```

twz123 commented on pkg/leaderelection/lease_pool_test.go:
```diff
@@ -198,7 +198,7 @@ func TestSecondWatcherAcquiresReleasedLease(t *testing.T) {
 	t.Log("Started first lease pool")
 
-	ctx2, cancel2 := context.WithCancel(context.TODO())
+	ctx2, cancel2 := context.WithCancel(t.Context())
 	t.Cleanup(cancel2)
 	events2, err := pool2.Watch(ctx2, withOutputChannels(&LeaseEvents{
```
```suggestion
	events2, err := pool2.Watch(t.Context(), withOutputChannels(&LeaseEvents{
```

The defer on line 208 is superfluous.

twz123 commented on pkg/leaderelection/lease_pool_test.go:
```diff
@@ -189,7 +189,7 @@ func TestSecondWatcherAcquiresReleasedLease(t *testing.T) {
 	t.Log("Pre-created acquired lease for first identity")
 
-	ctx1, cancel1 := context.WithCancel(context.TODO())
+	ctx1, cancel1 := context.WithCancel(t.Context())
 	t.Cleanup(cancel1)
```
```suggestion
	ctx1, cancel1 := context.WithCancel(t.Context())
	defer cancel1() // This is not required, but I can't silence the linter via nolint? :thinking:
```

twz123 commented on pkg/token/joinclient_test.go:
```diff
@@ -142,7 +143,7 @@ func TestJoinClient_Cancellation(t *testing.T) {
 
 func startFakeJoinServer(t *testing.T, handler func(http.ResponseWriter, *http.Request)) (*url.URL, []byte) {
-	requestCtx, cancelRequests := context.WithCancel(context.Background())
+	requestCtx, cancelRequests := context.WithCancel(t.Context())
```
```suggestion
```

And replace `BaseContext` on line 178 with `func(net.Listener) context.Context { return t.Context() },`

twz123 commented on pkg/token/joinclient_test.go:
```diff
@@ -181,7 +182,8 @@ func startFakeJoinServer(t *testing.T, handler func(http.ResponseWriter, *http.R
 	go func() { defer close(serverError); serverError <- server.ServeTLS(listener, "", "") }()
 	t.Cleanup(func() {
 		cancelRequests()
```
```suggestion
```

twz123 reviewed (commented):
We can bump the go directive in hack/tool/go.mod, too.

juanluisvaladas commented:
Thanks for the extensive review, everything is addressed

twz123 reviewed (commented):
Did you expand the "hidden conversations"? :upside_down_face:

juanluisvaladas commented on pkg/kubernetes/watch/watcher_test.go:
```diff
@@ -36,7 +36,7 @@ import (
 
 func TestWatcher(t *testing.T) {
```
Except `ContextInitiallyCanceled` and `InitialListError` which can't be removed, the rest use the context.WithTimeout defined at the beginning. I don't think there is anything to do. here.

The only thing I think we can remove is `t.Cleanup(cancel)` of line 40, which is redundant, but the linter complains if I remove that so it's either that or nolint directive.

juanluisvaladas commented:
I definitely missed the hidden conversations 🤦 It's ready to review again.
I addressed all your concerns, please review the comments on following files:
* pkg/kubernetes/watch/watcher_test.go
* pkg/component/worker/config/loader_test.go
* pkg/component/worker/nllb/reconciler_test.go

The rest is basically doing exactly what you pointed out so it's safe to skip.
